### PR TITLE
[Lint] Add a formatter GitHub Action

### PR DESCRIPTION
**Pull request by yaox12**
## Description

Leverage [lintrunner](https://github.com/pytorch/pytorch/wiki/lintrunner) to do format checks and suggestions for PRs via GitHub Action.

1. Only files modified in the current PR will be checked.
2. Use [UFMT](https://github.com/omnilib/ufmt), built on top of [black](https://black.readthedocs.io/) and [µsort](https://usort.readthedocs.io/) (an alternative to isort, see [here](https://usort.readthedocs.io/en/stable/why.html#comparison-to-isort) for comparison), to format Python code.
3. Use `clang-format` to format C/C++/CUDA code.

This action will post warnings/errors directly on the corresponding files :) See [here](https://github.com/yaox12/dgl/pull/57/files#diff-e22baf356d5f5ca22eb63e132344c6f3f92a94941ee5bc1ee3947929a6e5f89a) for example.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: .lintrunner.toml [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+]
+is_formatter = true
+
+[[linter]]
+code = 'CLANGFORMAT'
+include_patterns = [
+    '**/*.h',
+    '**/*.cc',
+    '**/*.cuh',
+    '**/*.cu',
```
Comment (frozenbugs) [inline .lintrunner.toml:38]: should we also include:
*.c
*.cpp

**Comment by frozenbugs** [comment]
I really like this!
But it seems that in the example, the warning only appears at the top of the file, and the warning is very generic.
Is it possible to have it the specific line? and call out the exact warning?

File: .github/workflows/lint.yml [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        continue-on-error: true
+        run: |
+          # Use jq to massage the JSON lint output into GitHub Actions workflow commands.
+          jq --raw-output \
+            '"::\(if .severity == "advice" or .severity == "disabled" then "warning" else .severity end) file=\(.path),line=\(.line),col=\(.char),title=\(.code) \(.name)::" + (.description | gsub("\\n"; "%0A"))' \
+            lint.json
+
+concurrency:
+  group: ${{ github.workflow }}-${{ github.event.pull_request.number || github.sha }}-${{ github.event_name == 'workflow_dispatch' }}
+  cancel-in-progress: true
```
[the remark this answers is not in the document]
Comment (yaox12) [reply .github/workflows/lint.yml]: Updated.

File: .lintrunner.toml (hunk 1 above)
Comment (yaox12) [reply .lintrunner.toml:40]: Updated.

**Comment by yaox12** [comment]
> I really like this! But it seems that in the example, the warning only appears at the top of the file, and the warning is very generic. Is it possible to have it the specific line? and call out the exact warning?

1. It should appear in the first line that needs to be changed now.
2. Since UFMT and clang-format are both formatters, it's hard to give exact warnings. Instead, they will give diffs before/after formatting. Click the upper right icon to see details. 
image

**Review by frozenbugs — APPROVED** [review]

**Pull request merged**